riscv-executor/src/lib.rs

```diff
@@ -645,59 +904,202 @@ struct Executor<'a, 'b, F: FieldElement> {
+            .unwrap()
+            .iter()
+            .enumerate()
+        {
+            if val.is_one() {
+                return i as u32;
+            }
+        }
+        panic!("could not find sink_id by looking at the p_instr__loop column");
+    }
+
+    /// read register value, updating the register memory machine
+    fn reg_read(&mut self, step_offset: u32, reg: u32, selector_idx: u32) -> Elem<F> {
+        let val = self.proc.get_reg_mem(reg);
+        self.proc
+            .regs_machine
+            .read(self.step + step_offset, reg, val, selector_idx);
+        val
+    }
+
+    /// write value to register, updating the register memory machine
+    fn reg_write(&mut self, step_offset: u32, reg: u32, val: Elem<F>, selector_idx: u32) {
+        if let ExecMode::Trace = self.mode {
+            self.proc
+                .regs_machine
+                .write(self.step + step_offset, reg, val, selector_idx);
+        }
+        self.proc.set_reg_mem(reg, val);
+    }
+
+    fn set_program_columns(&mut self, pc: u32) {
+        if let ExecMode::Trace = self.mode {
+            // set witness from the program definition
+            for (fixed, col) in &self.program_cols {
+                let val = Elem::Field(
+                    *self
+                        .get_fixed(fixed)
+                        .unwrap_or(&Vec::new())
+                        .get(pc as usize)
+                        .unwrap_or(&F::zero()),
+                );
+                self.proc.set_col(col, val);
+            }
+            // always 2 because RISCV only uses constrainted submachines
+            self.proc.set_col("main::_operation_id", 2.into());
+        }
+    }
+
     fn exec_instruction(&mut self, name: &str, args: &[Expression]) -> Vec<Elem<F>> {
+        // shorthand macros for setting/getting main machine witness values in the current row
+        macro_rules! set_col {
+            ($name:ident, $val:expr) => {
+                self.proc
+                    .set_col(concat!("main::", stringify!($name)), $val);
+            };
+        }
+        macro_rules! get_col {
+            ($name:ident) => {
+                self.proc.get_col(concat!("main::", stringify!($name)))
+            };
+        }
+
         let args = args
             .iter()
             .map(|expr| self.eval_expression(expr)[0])
             .collect::<Vec<_>>();
 
         self.proc.backup_reg_mem();
 
-        match name {
+        set_col!(X, get_col!(X_const));
+        set_col!(Y, get_col!(Y_const));
+        set_col!(Z, get_col!(Z_const));
+        set_col!(W, get_col!(W_const));
+        self.proc
+            .set_col(&format!("main::instr_{name}"), Elem::from_u32_as_fe(1));
+
+        let r = match name {
             "set_reg" => {
                 let addr = args[0].u();
-                self.proc.set_reg_mem(addr, args[1]);
+                let val = args[1];
+                self.reg_write(0, addr, val, 3);
```

Comment:
3?

Reply:
yep, that's the selector assigned to the link doing this "reg_memory.write"